components/public-api/gitpod/v1/workspaces.proto

```diff
@@ -0,0 +1,277 @@
+    rpc ListenToImageBuildLogs(ListenToImageBuildLogsRequest) returns (stream ListenToImageBuildLogsResponse) {}
+
+    // StopWorkspace stops a running workspace (instance).
+    // Errors:
+    //   NOT_FOUND:           the workspace_id is unkown
+    //   FAILED_PRECONDITION: if there's no running instance
+    rpc StopWorkspace(StopWorkspaceRequest) returns (stream StopWorkspaceResponse) {}
+
+}
+
+message ListWorkspacesRequest {
+    Pagination pagination = 1;
+    google.protobuf.FieldMask field_mask = 2;
+}
+message ListWorkspacesResponse {
+    // Status status = 1;
+    string next_page_token = 2;
+
+    message WorkspaceAndInstance {
+        Workspace result = 1;
+        WorkspaceInstance last_active_instances = 2;
+    }
+
+    repeated WorkspaceAndInstance result = 1;
+}
+
+message GetWorkspaceRequest {
+    string workspace_id = 1;
+}
+message GetWorkspaceResponse {
+    Workspace result = 1;
+}
+
+message CreateWorkspaceRequest {
+    string idempotency_token = 1;
+
+    string context_url = 2;
+
+    oneof prebuild {
+        // if true, and there's a prebuild available (not running), use that prebuild
+        bool if_available = 3;
+        // uses a particular prebuild - the prebuild ID is scoped by the context URL.
+        // Use the PrebuildService to get ahold of the prebuild_id.
+        string prebuild_id = 4;
+    }
+}
+message CreateWorkspaceResponse {
+    // Status status = 1;
+
+    string workspace_id = 2;
+}
+
+message StartWorkspaceRequest {
+    string idempotency_token = 1;
+
+    string workspace_id = 2;
+}
+message StartWorkspaceResponse {
+    // Status status = 1;
+
+    string instance_id = 2;
+    string workspace_url = 3;
+}
+
+message GetActiveWorkspaceInstanceRequest {
+    string workspace_id = 1;
+}
+message GetActiveWorkspaceInstanceResponse {
+    // Status status = 1;
+
+    WorkspaceInstance instance = 2;
+}
+
+message ListenToWorkspaceInstanceRequest {
+    string instance_id = 1;
+}
+message ListenToWorkspaceInstanceResponse {
+    // Status status = 1;
+
+    WorkspaceInstanceStatus instance_status = 2;
+}
+
+message ListenToImageBuildLogsRequest {
+    string instance_id = 1;
+}
+message ListenToImageBuildLogsResponse {
+    // Status status = 1;
+
+    string line = 2;
+}
+
+message StopWorkspaceRequest {
+    string idempotency_token = 1;
+
+    string workspace_id = 2;
+}
+message StopWorkspaceResponse {}
+
+////////////////////////////////
+// Shared messages come here
+////////////////////////////////
+
+// Workspace describes a single workspace
+message Workspace {
+    // workspace_id is the ID of the workspace
+    string workspace_id = 1;
+
+    // owner_id is the ID of the user who created this workspace
+    string owner_id = 2;
+
+    // project_id is the ID of the project which this workspace belongs to
+    string project_id = 3;
+
+    // context_url reports the original context URL the workspace was created from
+    string context_url = 4;
```

Comment:
Original means with prefixes? It would be nice to have context URL which can be also displayed in UI without prefixes. Dashboard right now has quite complicated logic to get such and Gateway UI we see null from time to time, since `normalizedContextUrl` is missing.

Reply:
we could just report the `normalized_context_url`. AFAIK there isn't a use-case for the original context URL, no?

Reply:
> we could just report the normalized_context_url.

It is a bit complicated, since not all workspaces have it and then we need to recompute normalized.

> AFAIK there isn't a use-case for the original context URL, no?

In IDE integrations I did not see any, not sure about dashboard @geropl ?

Reply:
Could we not "just" produce a normalised form for all contexts?

Reply:
> Could we not "just" produce a normalised form for all contexts?

And how to handle non-URl contexts like snapshots, then?

I have not find time to look at this PR in total, yet. But wrt `contextURL` what we learned in the past is that we need a clear separation between :
 - what was pasted into the browser URL field to start the workspace
 - (optional) a URL of the context, e.g. github.com/my/repo/tree/my-branch
 - a nicely displayable version of the context, e.g. "https://github.com/my/repo/tree/my-branch", or "Snapshot: xyz"

A single `string` field won't cut it I fear. Strongly voting for entity that we can extend as we evolve what a "Context" actually is.

Reply:
I've extended the `context_url` to a `context` type which can provide more details about the workspace context.